### PR TITLE
Fix the layout calculation in sliver list where the scroll offset cor…

### LINKED ISSUE #59819: AnimatedList/ListView missing elements on scroll
When updating a list in the middle of a scroll, animated list does not render all elements. Everything works as expected if the list is updated while scroll position is in the beginning of the list.

## Steps to Reproduce

https://dartpad.dev/5ea587e079e7a05f4954cfd680e3f71c

1. Scroll to middle/end of list.
2. Press 'Show' to unhide elements
3. Scroll up, list breaks, doesn't show all elements

**Expected results:** List should load all elements

**Actual results:** List stops loading at random.

### PR DESCRIPTION
**Pull request by chunhtai**
…rection accidently set the child with layout offset of zero

## Description

consider following example:
we have sliver list with items where each item has 10 paint extent. the screen is at item 10

-------start of screen----------
item 10 -- layout offset 100
item 11 -- layout offset 110
item 12 -- layout offset 120
item 13 -- layout offset 130
-------end of screen-----------

now, the sliver list children updated, and the each item has 20 paint extent. The RenderSliverList is smart to update the children scroll offset **after** the first child. So it become like this

-------start of screen----------
item 10 -- layout offset 100
item 11 -- layout offset 120
-------end of screen-----------

item 12 and 13 are not in the screen so it will be destroy.

Now, start scrolling back

-------start of screen----------
item 9 -- layout offset 80
item 11 -- layout offset 100
-------end of screen-----------

...keep scrolling

-------start of screen----------
item 5 -- layout offset 0
item 11 -- layout offset 20
-------end of screen-----------
oops, the sliver list think item 5 is the start of the list, so you cannot scroll any further.

## Related Issues

Fixes https://github.com/flutter/flutter/issues/59819

## Tests

I added the following tests:

see files

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [ ] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [ ] I signed the [CLA].
- [ ] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [ ] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] All existing and new tests are passing.
- [ ] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [ ] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [Handling breaking changes].

- [ ] No, no existing tests failed, so this is *not* a breaking change.
- [ ] Yes, this is a breaking change. *If not, delete the remainder of this section.*
   - [ ] I wrote a design doc: https://flutter.dev/go/template *Replace this with a link to your design doc's short link*
   - [ ] I got input from the developer relations team, specifically from: *Replace with the names of who gave advice*
   - [ ] I wrote a migration guide: https://flutter.dev/go/breaking-changes-template *Replace this with a link to a pull request that adds the migration guide to https://flutter.dev/docs/release/breaking-changes*

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

File: packages/flutter/test/widgets/slivers_test.dart [hunk 1]
```diff
@@ -308,6 +308,107 @@ void main() {
     },
   );
 
+  testWidgets(
+    'SliverList can handle inaccurate scroll offset due to changes in children list',
+      (WidgetTester tester) async {
+      bool skip = true;
```
Comment (Piinks) [inline packages/flutter/test/widgets/slivers_test.dart:314]: Can you add one of those issue references in case it gets lost in history?

**Review by Piinks — APPROVED** [review]
NICE! 🎉 

Flutter_LGTM

File: packages/flutter/test/widgets/slivers_test.dart [hunk 2]
```diff
@@ -308,6 +308,107 @@ void main() {
     },
   );
 
+  testWidgets(
```
Comment (goderbauer) [inline packages/flutter/test/widgets/slivers_test.dart:311]: Add: "Regression test for <issue link>" here?

File: packages/flutter/lib/src/rendering/sliver_list.dart [hunk 3]
```diff
@@ -180,6 +164,28 @@ class RenderSliverList extends RenderSliverMultiBoxAdaptor {
       trailingChildWithLayout ??= earliestUsefulChild;
     }
 
+    assert(childScrollOffset(firstChild) > -precisionErrorTolerance);
+
+    // If the scroll offset is at the zero, we should make sure we are
```
Comment (goderbauer) [inline packages/flutter/lib/src/rendering/sliver_list.dart:169]:
```suggestion
    // If the scroll offset is at zero, we should make sure we are
```

File: packages/flutter/lib/src/rendering/sliver_list.dart [hunk 4]
```diff
@@ -180,6 +164,28 @@ class RenderSliverList extends RenderSliverMultiBoxAdaptor {
...
+    // actually at the beginning of the list
```
Comment (goderbauer) [inline packages/flutter/lib/src/rendering/sliver_list.dart:170]: nit: end with a `.`

**Review by goderbauer — APPROVED** [review]
LGTM

**Comment by flutter-github-sync** [comment]
Google testing passed!

**Comment by flutter-github-sync** [comment]
Google testing passed!

**Pull request merged**